### PR TITLE
Location protos

### PR DESCRIPTION
**Pull request by RyanLassigBanks**
Add protobuf definitions and structs for implementing the locations RFC

### CODE REVIEW

File: sdk/protos/location_payload.proto [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    string owner = 3;
+    repeated PropertyValue properties = 4;
+}
+
+message LocationUpdateAction {
+    // Not modified. Only  used to find location object in state
+    Location.LocationType location_type = 1;
+    // Not modified. Only  used to find location object in state
+    string location_id = 2;
+      // This will replace all properties currently defined
```
Comment (agunde406) [inline sdk/protos/location_payload.proto:51]: extra two spaces here

File: sdk/protos/location_state.proto [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import "schema_state.proto";
+
+message Location {
+    // Who owns this product (pike organization id)
+    enum LocationType {
+      UNSET_TYPE = 0;
+      GS1 = 1;
+    }
+
+    //  Global Location Number as defined by GS1 specification
```
Comment (agunde406) [inline sdk/protos/location_state.proto:26]: extra space

File: sdk/protos/location_payload.proto (hunk 1 above)
Comment (RyanLassigBanks) [reply sdk/protos/location_payload.proto]: Fixed

File: sdk/protos/location_state.proto (hunk 2 above)
Comment (RyanLassigBanks) [reply sdk/protos/location_state.proto:25]: Fixed

Comment (agunde406) [reply sdk/protos/location_state.proto:25]: Sorry, the one I was referring to was after the //

Comment (RyanLassigBanks) [reply sdk/protos/location_state.proto]: Ah I see. Updated.

File: sdk/protos/location_payload.proto [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    // Not modified. Only  used to find location object in state
+    Location.LocationType location_type = 1;
+    // Not modified. Only  used to find location object in state
+    string location_id = 2;
+    // This will replace all properties currently defined
+    repeated PropertyValue properties = 3;
+}
+
+message LocationDeleteAction {
+    // Not modified. Only  used to find location object in state
```
Comment (agunde406) [inline sdk/protos/location_payload.proto:56]: Probably don't need the "not modified" here as deleting is modifying. Instead just say "Used to find location object in state" for both

File: sdk/src/protocol/location/state.rs [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    }
+}
+
+impl IntoProto<protos::location_state::Location> for Location {}
+impl IntoNative<Location> for protos::location_state::Location {}
+
+#[derive(Debug)]
+pub enum LocationBuildError {
+    MissingField(String),
+    EmptyVec(String),
```
Comment (agunde406) [inline sdk/src/protocol/location/state.rs:155]: EmptyVec is never used

File: sdk/src/protocol/location/state.rs [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+        let location_type = self.location_type.ok_or_else(|| {
+            LocationBuildError::MissingField("'location_type' field is required".to_string())
+        })?;
+
+        let owner = self.owner.ok_or_else(|| {
+            LocationBuildError::MissingField("'owner' field is required".to_string())
+        })?;
+
+        // Product values are not required
```
Comment (agunde406) [inline sdk/src/protocol/location/state.rs:229]: Product values?

File: sdk/src/protocol/location/state.rs [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+
+    pub fn build(self) -> Result<LocationList, LocationBuildError> {
+        // Product values are not required
+        let locations = self.locations.ok_or_else(|| {
+            LocationBuildError::MissingField("'locations' field is required".to_string())
+        })?;
+
+        let locations = {
+            if locations.is_empty() {
+                return Err(LocationBuildError::MissingField(
```
Comment (agunde406) [inline sdk/src/protocol/location/state.rs:342]: Should this be EmptyVec?

File: sdk/src/protocol/location/state.rs [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+    pub fn build(self) -> Result<LocationList, LocationBuildError> {
+        // Product values are not required
+        let locations = self.locations.ok_or_else(|| {
+            LocationBuildError::MissingField("'locations' field is required".to_string())
+        })?;
+
+        let locations = {
+            if locations.is_empty() {
+                return Err(LocationBuildError::EmptyVec(
+                    "'locations' cannot be empty".to_string(),
```
Comment (agunde406) [inline sdk/src/protocol/location/state.rs:342]: This is going to end up saying "'locations' cannot be empty must not be empty", this should be changed to just 'locations'

File: sdk/protos/location_payload.proto [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    // timestamp
+    uint64 timestamp = 2;
+
+    LocationCreateAction location_create = 3;
+    LocationUpdateAction location_update = 4;
+    LocationDeleteAction location_delete = 5;
+}
+
+message LocationCreateAction {
+    Location.LocationNamespace namespace = 1;
```
[the remark this answers is not in the document]
Comment (RyanLassigBanks) [reply sdk/protos/location_payload.proto]: Moved namespace protos into `location_namespace.proto`

File: sdk/src/protocol/location/payload.rs [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+// limitations under the License.
+
+use protobuf::Message;
+use protobuf::RepeatedField;
+
+use std::error::Error as StdError;
+
+use super::errors::BuilderError;
+
+use crate::protocol::{location::state::LocationNamespace, schema::state::PropertyValue};
```
[the remark this answers is not in the document]
Comment (RyanLassigBanks) [reply sdk/src/protocol/location/payload.rs]: Move LocationNamespace enum into `location/namespace.rs`

File: sdk/protos/location_namespace.proto [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+syntax = "proto3";
+
+enum LocationNamespace {
```
Comment (vaporos) [inline sdk/protos/location_namespace.proto:18]: Making other protos include this on this ties them together unnecessarily.

**Pull request merged**